Implement WebXR Interstitial

MortimerGoro opened the pull request:
Fixes  #1966 Fixes #3123

MortimerGoro commented:
The spinner animation is not the final one, @keianhzo is working on the shapeshifter exported version

daoshengmu commented:
There is a build fail on this merge:

```
/opt/FirefoxReality/app/src/common/shared/org/mozilla/vrbrowser/ui/widgets/WebXRInterstitialController.java:15: error: cannot find symbol
    private WebxrInterstitialControllerBinding mBinding;
```

MortimerGoro commented:
> There is a build fail on this merge:
> 
> ```
> /opt/FirefoxReality/app/src/common/shared/org/mozilla/vrbrowser/ui/widgets/WebXRInterstitialController.java:15: error: cannot find symbol
>     private WebxrInterstitialControllerBinding mBinding;
> ```

Fixed

daoshengmu commented:
I notice the controller icons only be shown when entering the immersive mode at the 1st time. Is it what we want?

MortimerGoro commented:
For now it’s displayed for 3 seconds the first time you enter immersive (on each app restart.
I don’t like forcing it every EnterVR.

After trying it I think it would be good to wait for a user click to dismiss the how-to information, and remove the forced time. IMO:
- The first time you go to WebXR show the how-to-exit information. Dismiss after a click, not time based. We don’t know how long takes for a user to process the info and find the button.
- Do not show the spinner at the same time, it distracts from reading the how-to
- Show the spinner only in the next EnterVR for the whole session

Thoughts?

bluemarvin commented:
> For now it’s displayed for 3 seconds the first time you enter immersive (on each app restart)
> I don’t like forcing it every EnterVR.
> 
> After trying it I think it would be good to wait for a user click to dismiss the how-to information, and remove the forced time. IMO:
> 
>     * The first time you go to WebXR show the how-to-exit information. Dismiss after a click, not time based. We don’t know how long takes for a user to process the info and find the button.
> 
>     * Do not show the spinner at the same time, it distracts from reading the how-to
> 
>     * Show the spinner only in the next EnterVR for the whole session
> 
> 
> Thoughts?

I think this is a good approach. I think the controller images could be a little bigger maybe?

daoshengmu commented:
> For now it’s displayed for 3 seconds the first time you enter immersive (on each app restart)
> I don’t like forcing it every EnterVR.
> 
> After trying it I think it would be good to wait for a user click to dismiss the how-to information, and remove the forced time. IMO:
> 
> * The first time you go to WebXR show the how-to-exit information. Dismiss after a click, not time based. We don’t know how long takes for a user to process the info and find the button.
> * Do not show the spinner at the same time, it distracts from reading the how-to
> * Show the spinner only in the next EnterVR for the whole session
> 
> Thoughts?

I feel we can remove the spinner if we are showing the how-to-exit information, the spinner sometimes catch my attention than the how-to-exit. However, if users end the information icon, WebXR pages are still loading. We should still show the spinner instead.

Make the controller image bigger can let me see the words more easy.  I also feel if we already allow users to press a button to end the how-to-exit information, maybe longer time like 5 secs would be great for user have more time to understand it.

bluemarvin commented on app/src/common/shared/org/mozilla/vrbrowser/utils/DeviceType.java:
```diff
@@ -2,25 +2,48 @@
+            case OculusQuest:
+                name = "Oculus Quest";
+                break;
+            case ViveFocusPlus:
+                name = "Vive Focus Plus";
+                break;
```
Do we need a `ViveFocus` also?

bluemarvin commented on app/src/main/cpp/BrowserWorld.cpp:
```diff
@@ -1481,20 +1494,22 @@ BrowserWorld::DrawImmersive(device::Eye aEye) {
+  m.rootWebXRInterstitial->SetTransform(m.device->GetReorientTransform());
   m.drawHandler = [=](device::Eye eye) {
-    DrawLoadingAnimation(eye);
+      DrawWebXRInterstitial(eye);
```
nit: Looks like there are two extra spaces added?

bluemarvin commented on app/src/common/shared/org/mozilla/vrbrowser/utils/DeviceType.java:
```diff
@@ -2,25 +2,48 @@
 
 public class DeviceType {
     // These values need to match those in Device.h
+    @IntDef(value = { Unknown, OculusGo, OculusQuest, ViveFocusPlus, PicoNeo2, PicoG2 })
```
It looks like `Device.h` needs to be updated?

daoshengmu replied:
We need ViveFocus as well

daoshengmu commented:
Do we have an alpha blending problem on the spinner image? I can see its rectangle background when we are going to end the animation.

MortimerGoro commented:
@daoshengmu @bluemarvin I implemented the flow we talked about. Let me know what you think. Feel free to change the new text

daoshengmu commented:
> @daoshengmu @bluemarvin I implemented the flow we talked about. Let me know what you think. Feel free to change the new text

It seems like pressing squeeze btns on Quest can't exit the how-to-exit page.

MortimerGoro commented:
> > @daoshengmu @bluemarvin I implemented the flow we talked about. Let me know what you think. Feel free to change the new text
> 
> It seems like pressing squeeze btns on Quest can't exit the how-to-exit page.

Fixed

MortimerGoro commented:
> Do we have an alpha blending problem on the spinner image? I can see its rectangle background when we are going to end the animation.

I noticed that too but it seems a driver bug. It only happens with the first layer displayed on a black background. If I change the order the alpha blending moves to the first layer rendered. I'll look into it next week, we can address that as a follow-up

daoshengmu commented on app/src/oculusvr/cpp/DeviceDelegateOculusVR.cpp:
```diff
@@ -483,10 +483,6 @@ struct DeviceDelegateOculusVR::State {
-        // This is always false in Oculus Browser.
-        const bool thumbRest = false;
-        controller->SetButtonState(controllerState.index, ControllerDelegate::BUTTON_OTHERS, 5, thumbRest, thumbRest);
       } else {
```
It looks odd to me. This is thumbRest instead of the grip button, and thumbRest is always true. If `thumbRest` will cause problems to you, Neo2 will have it too because it also has `thumbRest`.

daoshengmu replied:
You can remove this commit now. I already gave a fix at #2521 by defining a new btn state `BUTTON_SQUEEZE`. That makes sense for us to use it in app.
